File: src/main/java/org/stt/validation/ItemAndDateValidator.java
package org.stt.validation;

import org.stt.query.Criteria;
import org.stt.query.TimeTrackingItemQueries;
import org.stt.time.DateTimes;
import org.stt.time.Interval;

import javax.inject.Inject;
import java.time.LocalDateTime;

import static java.util.Objects.requireNonNull;

public class ItemAndDateValidator {
    private final TimeTrackingItemQueries timeTrackingItemQueries;

    @Inject
    public ItemAndDateValidator(TimeTrackingItemQueries timeTrackingItemQueries) {
        this.timeTrackingItemQueries = requireNonNull(timeTrackingItemQueries);
    }

    public boolean validateItemIsFirstItemAndLater(LocalDateTime start) {
        if (!DateTimes.isToday(start)) {
            return true;
        }
        LocalDateTime startOfDay = start.toLocalDate().atStartOfDay();
        Interval searchInterval = Interval.between(startOfDay, start);
        Criteria criteria = new Criteria();
        criteria.withStartBetween(searchInterval);
        boolean hasEarlierItem = timeTrackingItemQueries.queryItems(criteria)
                .findAny()
                .isPresent();
        if (hasEarlierItem) {
            return true;
        }
        return !LocalDateTime.now().isBefore(start);
    }
}

File: src/main/kotlin/org/stt/validation/ItemAndDateValidator.kt
package org.stt.validation

import org.stt.query.Criteria
import org.stt.query.TimeTrackingItemQueries
import org.stt.time.DateTimes
import org.stt.time.until
import java.time.LocalDateTime
import javax.inject.Inject

class ItemAndDateValidator @Inject
constructor(private val timeTrackingItemQueries: TimeTrackingItemQueries) {
    fun validateItemIsFirstItemAndLater(start: LocalDateTime): Boolean {
        if (!DateTimes.isToday(start)) {
            return true
        }
        val startOfDay = start.toLocalDate().atStartOfDay()
        val searchInterval = startOfDay until start
        val criteria = Criteria()
        criteria.withStartBetween(searchInterval)
        val hasEarlierItem = timeTrackingItemQueries.queryItems(criteria)
                .findAny()
                .isPresent
        return hasEarlierItem || !LocalDateTime.now().isBefore(start)
    }
}
